### PR TITLE
Optimization#7 Parallel execution of JUnit 4/5 tests

### PR DESCRIPTION
**Pull request by jesus-gorronogoitia**
This PR replaces the #775 which is corrupted.

### CODE REVIEW

**Comment by jesus-gorronogoitia** [comment]
Hi @danglotb 
I've created this new PR replacing #775 
Previous one got corrupted for reasons I don't understand, but maybe I did not rebase well the master changes (which were frequent during this PR preparation) so I got lots of conflicts during the rebase that were corrupting the code.

This is the procedure for rebasing I was following. Please, let me know if it correct:
I have two remotes, one to my forked DSpot (origin), another to the original DSpot (upstream):
```
yosu@avalon:~/Projects/STAMP/Git/dspot-yosu/dspot> git remote -v
origin	https://github.com/jesus-gorronogoitia/dspot.git (fetch)
origin	https://github.com/jesus-gorronogoitia/dspot.git (push)
upstream	https://github.com/STAMP-project/dspot.git (fetch)
upstream	https://github.com/STAMP-project/dspot.git (push)
```
So, then, I switch to master branch:
`git checkout master`

Then, I merge upstream changes:
`git pull upstream master`

Then, I switch to my branch:
`git checkout optimization#7`

And finally, I rebase master:
`git rebase master`

Is this process correct?
Typically, when I rebase master, I get may conflicts for all the intermediate commits between master and optimization#7

**Comment by jesus-gorronogoitia** [comment]
Hi @danglotb 
Compilation if failing in DSpot Maven plugin because it has an strong dependency in DSpot configuration. In my opinion, the way DSpot Configuration configuration method is defined (with a long list of configurations) is not the best one, as if someone (like me) add a new configuration flag in DSpot, and therefore, in this method, it breaks the other plugins. I am not familiarized with DSpot Maven plugin as to easily fix changes in DSpot configuration.
What do you think we should do?
BTW, the way parallel test execution is configure uses only one flag:
--execute-test-parallel-with-number-processors <number>
if this flag is set, two DSpot configurations are created, one exactly as this one, another one, boolean, saying tests should be executed in parallel. If above number is 0, then the number of processors is taken from the default CPU #processors.
In other words, configuring Maven DSpot could be a little tricky, but I prefer to use only one flag and setup internally the two flags required.
Please, let me know how can we fix this PR issues, so I can be accessed before DSpot is evolving with new commits in master that can compromise this PR if I am not rebasing it correctly.

**Comment by danglotb** [comment]
Hi @jesus-gorronogoitia, 

Thank you for doing this.

The rebase procedure is correct. But what do you do when you meet a conflict?

After resolving the conflict (by editing the files and removing the `HEAD <<<<<<` and `>>>>>>> commit_msg'), you should:

```
git add <fileInConflict>
git rebase --continue
```

if you want to list the current files that have conflicts, you can run a ```git status``` and files under the category `both modified`.

I understand your point about the configuration. However, what would be the best practice? The maven-plugin replicates the command line options as maven parameters.

To make updating the maven parameters simple, quick and to ensure it, I made recent changes:

You have a method that generates the textual code of the parameters (and the documentation) of the DSpotMojo, see [JSAPOptions#main()](https://github.com/STAMP-project/dspot/blob/master/dspot/src/main/java/eu/stamp_project/utils/options/JSAPOptions.java#L425).

You will have just to copy and paste the text of the standard output in the [DSpotMojo](https://github.com/STAMP-project/dspot/blob/master/dspot-maven/src/main/java/eu/stamp_project/DSpotMojo.java#L27) by replacing the current parameters, _i.e._ form `@Parameter(property = "path-to-properties")`[Line 27](https://github.com/STAMP-project/dspot/blob/master/dspot-maven/src/main/java/eu/stamp_project/DSpotMojo.java#L27) to `@Parameter(defaultValue = "false", property = "help")`[Line 224](https://github.com/STAMP-project/dspot/blob/master/dspot-maven/src/main/java/eu/stamp_project/DSpotMojo.java#L224).

Others parameters are specific to `dspot-maven` (but could(should?) be moved to `dspot-core`).

Then, you have the update the call to [Configuration.configure()](https://github.com/STAMP-project/dspot/blob/master/dspot-maven/src/main/java/eu/stamp_project/DSpotMojo.java#L276) which is exactly the same call done in [JSAPOptions#parse()](https://github.com/STAMP-project/dspot/blob/master/dspot/src/main/java/eu/stamp_project/utils/options/JSAPOptions.java#L82).

This is the smallest and simplest change that: ensure that `dspot-maven` is kept up to date along the `dspot-core` that I could find. Actually, the compilation error is the way to ensure that the `dspot-maven`'s parameters are kept updated.

For me, this is a good way to verify that we keep the maven plugin up to date with the new options and typically not leave behind it from the `dspot-core`, and it is almost automatic.

I would like to make it simpler, quicker or most of all completely automatic. This could be done by using `Spoon` but I kind running out of time for this kind of stuff which is an extra feature for the comfort and 
 pure software engineering.

If you do not manage to update `dspot-maven`, please tell me and I'll open a pull-request on your pull-request to fix it.

Thank you.

**Comment by jesus-gorronogoitia** [comment]
Hi @danglotb 
I've fixed dspot-maven configuration for new DSpot flag on parallel test execution, so all DSpot plugins compile well. Please, check if my proposed changes are fine with you and aligned to the changes you sent me recently in your PR. Sorry that we have been working in parallel in the same stuff.

**Comment by jesus-gorronogoitia** [comment]
Hi @danglotb 
TravisCI reports tests errors, but I can't understand why there are test errors, since I added a new feature (test parallel execution) that is not triggered by existing tests. Indeed in some cases, tests are failing because the unavailability on disk of some test classes. What could I do? Any hint?

**Comment by danglotb** [comment]
Hello @jesus-gorronogoitia 

It is not a big deal for the changes.

For the errors, I can't tell right now. I'm looking.

**Comment by jesus-gorronogoitia** [comment]
Hi @danglotb 
"It is not a big deal for the changes." ??? Do you want me to accept your PR o my last changes to fix Dspot Maven are fine?

**Comment by danglotb** [comment]
I meat that I was not a big deal that we work both on the same thing.

We will integrate your changes after we fix the tests. I'm gonna close my PR on your fork.

Sorry, that was not that clear :)

The good news is that I can have the failures locally. Debugging is in progress.

File: dspot/src/main/java/eu/stamp_project/automaticbuilder/maven/MavenAutomaticBuilder.java [hunk 1]
```diff
@@ -66,6 +66,22 @@ public String compileAndBuildClasspath() {
         return this.classpath;
     }
 
+    private int computeClasspath(String... goals) {
+        final String pomPathname;
+        if (InputConfiguration.get().shouldExecuteTestsInParallel()) {
+            DSpotPOMCreator.createNewPomForComputingClassPathWithParallelExecution();
+            pomPathname = InputConfiguration.get().getAbsolutePathToProjectRoot()
+                    + DSpotPOMCreator.getParallelPOMName();
+        } else {
```
Comment (danglotb) [inline dspot/src/main/java/eu/stamp_project/automaticbuilder/maven/MavenAutomaticBuilder.java]: The else branch is wrong. This method is responsible for:

1. Resolve the dependencies
2. Build a file that contains all the dependencies.
3. Compile the classes, test classes and resources to have a proper `target` folder.

In fact here, we must use the default `pom.xml` for several reasons:

1. The generated pom by `DSpot` detaches the `testCompile` goal from any phase. This is done to be able to execute the instrumented tests through maven goals during the assertion amplification process:
```xml
            <plugin>
                <groupId>org.apache.maven.plugins</groupId>
                <artifactId>maven-compiler-plugin</artifactId>
                <executions>
                    <execution>
                        <id>default-testCompile</id>
                        <phase>none</phase>
                    </execution>
                </executions>
            </plugin>
```
Otherwise, maven would detect a difference between the sources and the current binary. Maven would recompile the test classes and thus we will the instrumentation making impossible to complete the assertion amplification process.

2. The users might have a specific way to build its test classes (don't remember the example but I know that encounter this issue).

I opened a new pull request with a fix.

**Comment by jesus-gorronogoitia** [comment]
Hi @danglotb 
I've accepted your PR, let's see if TravisCI gives now green light.

**Comment by jesus-gorronogoitia** [comment]
Hi @danglotb every thing went fine excepting the coverage decrease. Please, let me know if this is fine with you or if I need to so something else.

**Comment by jesus-gorronogoitia** [comment]
Hi @danglotb 
Everything went fine excepting the coverage decrease. Let me know if this is fine to accept the PR or if I should do something else.
